### PR TITLE
[back] feat: enable non-strict option in unconnected entities API route

### PR DESCRIPTION
**Pull request by lfaucon**
This is a first step for #1247. Next step would be to use this option in the frontend entity selector, but I prefer making this change once the backend route is approved.

### CODE REVIEW

File: backend/tournesol/views/unconnected_entities.py [hunk 1]
```diff
@@ -75,7 +87,8 @@ def get_related_entities(entity_id):
         )
 
         return (
-            Entity.objects.filter(id__in=user_all_entities - user_related_entities)
+            Entity.objects.filter(id__in=entity_ids_filter)
+            .annotate(distance=Case(*distance_annotation_whens, default=1+len(neighbors)))
             .annotate(n_comparisons=Subquery(comparison_counts))
-            .order_by("n_comparisons")
+            .order_by("-distance", "n_comparisons")
```
Comment (amatissart) [inline backend/tournesol/views/unconnected_entities.py:93]: The performance of this queryset seems to be okay actually. The bad performance we observed was probably due to heavy/unoptimized test setup :shrug: 

Locally on the dev-env with 1277 compared videos, I get a response from the API in ~200ms.

To be extra cautious, we could filter this queryset to keep (e.g) only the 1000 entities with the largest distance. That would prevent the SQL query from becoming very large (and possibly very slow) for some users in the future, without the need to re-implement the pagination.

File: backend/tournesol/views/unconnected_entities.py [hunk 2]
```diff
@@ -32,38 +46,36 @@ class UnconnectedEntitiesView(PollScopedViewMixin, generics.ListAPIView):
...
+                    distances[next_node] = distance+1
+                    bfs_queue.append((next_node, distance+1))
 
-        user_related_entities = get_related_entities(source_node.id)
-        user_all_entities = set()
+        # The entities filtered are the entities which don't have a distance
+        # because they are unreachable if the strict option has been left to
+        # True (the default). Otherwise all the entities except the source are
+        # selected.
+        strict = self.request.GET.get("strict") != "false"
```
[the remark this answers is not in the document]
Comment (lfaucon) [reply backend/tournesol/views/unconnected_entities.py]: Done in new version

File: backend/tournesol/views/unconnected_entities.py [hunk 3]
```diff
@@ -18,7 +19,20 @@
     get=extend_schema(
         description="List unconnected entities of the current user from a target entity"
                     " and the user's graph of comparisons (entities are vertices and"
-                    " comparisons are edges)."
+                    " comparisons are edges).",
+        parameters=[
+            OpenApiParameter(
+                "strict",
+                type=OpenApiTypes.STR,
```
[the remark this answers is not in the document]
Comment (lfaucon) [reply backend/tournesol/views/unconnected_entities.py]: Thanks! Changed in new version

File: backend/tournesol/views/unconnected_entities.py (hunk 1 above)
Comment (lfaucon) [reply backend/tournesol/views/unconnected_entities.py]: I went on an optimized it anyway. Customizing the Paginator class was not too difficult. I also added tests that cheeck that the API call takes < 1s.

Comment (lfaucon) [reply backend/tournesol/views/unconnected_entities.py]: Even the 100'000 videos with ~200'000 comparisons passes the test for < 1s api call (although the setup for the test takes 10 minutes (that's surprisingly slow...)

File: backend/tournesol/views/unconnected_entities.py [hunk 4]
```diff
@@ -28,54 +58,40 @@ class UnconnectedEntitiesView(PollScopedViewMixin, generics.ListAPIView):
...
+        # True (the default). Otherwise all the entities except the source are
+        # selected.
+        strict = self.request.query_params.get("strict") != "false"
+        entity_ids = set(e_id for e_id in neighbors if (
+            (e_id not in distances) if strict else (distances.get(e_id, 2) > 1)
+        ))
+        sorted_entity_ids = sorted(
+            entity_ids,
+            # Sorting first by distance and secondly by number of comparisons
+            key=lambda x: (-distances.get(x, math.inf), len(neighbors.get(x, [])))
```
Comment (amatissart) [inline backend/tournesol/views/unconnected_entities.py:95]: To ensure the sorted list is deterministic when requesting multiple pages:
```suggestion
            key=lambda x: (-distances.get(x, math.inf), len(neighbors.get(x, []), x))
```

**Review by amatissart — APPROVED** [review]

File: backend/tournesol/views/unconnected_entities.py (hunk 4 above)
Comment (lfaucon) [reply backend/tournesol/views/unconnected_entities.py]: Fixed. Thanks!

**Pull request merged**